sns_ik_lib/src/sns_acc_ik_base.cpp

```diff
@@ -216,6 +216,108 @@ SnsIkBase::ExitCode SnsAccIkBase::solve(const Eigen::MatrixXd& J, const Eigen::V
   return ExitCode::InternalError;
 }
 
+/*************************************************************************************************/
+
+SnsIkBase::ExitCode SnsAccIkBase::solve(const Eigen::MatrixXd& J, const Eigen::VectorXd& dJdq,
+                                        const Eigen::VectorXd& ddx, const Eigen::VectorXd& ddqCS,
+                                        Eigen::VectorXd* ddq, double* taskScale, double* taskScaleCS)
+{
+  // Input validation
+  if (size_t(ddqCS.rows()) != getNrOfJoints()) {
+    ROS_ERROR("Bad Input: ddqCS.rows() == nJnt is required!");
+    return ExitCode::BadUserInput;
+  }
+
+  //--- get the solution for the primary goal
+  Eigen::VectorXd ddq1;
+  SnsIkBase::ExitCode exitCode = solve(J, dJdq, ddx, &ddq1, taskScale);
+  if (exitCode != ExitCode::Success) {
+    ROS_ERROR("Primary goal did not find a solution! Terminating..");
+    return exitCode;
+  }
+
+  //--- find the solution for the secondary goal
+
+  *taskScaleCS = 1.0;  // task scale (assume feasible solution until proven otherwise)
+  Eigen::MatrixXd I = Eigen::MatrixXd::Identity(getNrOfJoints(), getNrOfJoints());
+
+  // Keep track of which joints are saturated:
+  std::vector<bool> jointIsFree(getNrOfJoints(), true);
+
+  /*
+   * W is a diagonal selection matrix which indicates free joints.
+   * The entry of 1 indicates the corresponding joint is free for the task,
+   * and 0 indicates the corresponding joint is saturated.
+   */
+  Eigen::MatrixXd W = I;  // null-space selection matrix
+  for (size_t jntIdx = 0; jntIdx < getNrOfJoints(); jntIdx++) {
+    if (ddq1(jntIdx) > (getUpperBounds())(jntIdx) + BOUND_TOLERANCE ||
+        ddq1(jntIdx) < (getLowerBounds())(jntIdx) - BOUND_TOLERANCE) {
+          W(jntIdx, jntIdx) = 0.0;
+          jointIsFree[jntIdx] = false;
+    }
+  }
+
+  // Compute nullspace projection matrix
+  Eigen::MatrixXd Jinv, Pinv;
+  if(!pinv(J, &Jinv, LIN_SOLVE_RESIDUAL_TOL)) {
+    ROS_ERROR("Pseudo-inverse of J cannot be computed!");
+    return ExitCode::InternalError;
+  }
+
+  Eigen::MatrixXd P1 = I - Jinv*J; // for primary task
+  if(!pinv((I - W)*P1, &Pinv, LIN_SOLVE_RESIDUAL_TOL)){
+    // if (I-W) is a zero matrix, inverse is the same
+    Pinv = (I - W)*P1;
+  }
+  Eigen::MatrixXd Pcs = (I - Pinv)*P1; // for both primary and joint saturation tasks
+
+  // Compute "a" and "b" from the paper.   (J*W*a = dx)
+  Eigen::VectorXd a = Pcs * ddqCS;
+  Eigen::ArrayXd b = ddq1.array();
+
+  // Compute the task scale associated with each joint
+  Eigen::ArrayXd jntScaleFactorArr(getNrOfJoints());
+  Eigen::ArrayXd lowMargin = (getLowerBounds() - b);
+  Eigen::ArrayXd uppMargin = (getUpperBounds() - b);
```

Comment:
should this be getUpperBounds() + b ?

Reply:
-b is correct here.

Reply:
worth a comment? Unless this is obvious to people with more familiarity

Reply:
Sounds good. I will add more comments here.